librz/util/base16.c

```diff
@@ -0,0 +1,290 @@
+	rz_return_val_if_fail(src, NULL);
+	size_t buf_sz = calculate_dest_length(n) + 1;
+	char *out = (char *)malloc(buf_sz);
+	if (!out) {
+		return NULL;
+	}
+	(void)rz_base16_encode(out, src, n);
+	return out;
+}
+
+/**
+ * \brief Decode a Base16-encoded message.
+ * \param[out] dest The decoded output.
+ * \param[in] src The Base16-encoded message.
+ * \param n The length of the encoded message.
+ * \return The length of the decoded message, excluding the NUL byte.
+ * \attention The \p dest parameter must have sufficient space to
+ * accommodate the decoded output, including the NUL byte.
+ * Specifically, at least \c 1 + n/2 bytes must be available.
+ *
+ * Decode a Base16-encoded message. If \p n is negative, the length of
+ * \p src is determined via strlen. The decoded output is stored in \p
+ * dest and is NUL byte terminated.
+ *
+ * If either \p dest or \p src is \c NULL, nothing is done and the
+ * value \c 0 is returned.
+ *
+ * Returns \c -1 if:
+ * - Characters in \p src are invalid.
+ * - \p n was \c -1 and the length of \p src exceeds \c ST64_MAX.
+ * - \p n is not an even number (invalid Base16 length).
+ *
+ * # Example
+ * \code{.c}
+ * const char *b16 = "4869"; // "Hi"
+ * ut8 buf[16];
+ * st64 len = rz_base16_decode(buf, b16, -1);
+ * assert(len == 2);
+ * assert(memcmp(buf, "Hi", 2) == 0);
+ * \endcode
+ */
+RZ_API st64 rz_base16_decode(RZ_OUT RZ_NULLABLE ut8 *dest, RZ_NULLABLE const char *src, st64 n) {
+	char buf[2];
+	ut8 tmp[1];
+	size_t i, j;
+	st64 ret = 0;
+
+	rz_return_val_if_fail(src, 0);
+	rz_return_val_if_fail(dest, 0);
+
+	n = calculate_src_length(src, n);
+	if (n == -1) {
+		return -1;
+	}
+
+	if (n % 2 != 0) {
+		return -1;
+	}
+
+	for (i = j = 0; i < (size_t)n; i++) {
+		int c = src[i];
+		if (is_base16(c)) {
+			buf[j++] = c;
+			if (j == 2) {
+				unpack_from4(tmp, (const ut8 *)buf);
+				*dest++ = tmp[0];
+				ret += 1;
+				j = 0;
+			}
+		} else {
+			return -1;
+		}
+	}
+
+	*dest = '\0';
+	return ret;
+}
+
+/**
+ * \brief Dynamically decode a Base16‑encoded string.
+ * \param[in]  src  Pointer to the Base16 text.
+ * \param      len  The length of \p src; pass \c -1 to use \c strlen(src).
+ * \return A newly allocated buffer holding the decoded binary data and
+ *         terminated with a NUL byte, or \c NULL on error.
+ *
+ * Memory is obtained with \c malloc and must be released with \c free
+ * by the caller.  The function returns \c NULL when:
+ *  - \p src is \c NULL,
+ *  - the string length is odd or exceeds \c ST64_MAX,
+ *  - a memory allocation fails, or
+ *  - the decoder encounters malformed characters.
+ *
+ * The decoded binary size of a Base16 string of length \p len is
+ * \c len/2 bytes, plus one byte for the NUL terminator.
+ */
+RZ_API RZ_OWN ut8 *rz_base16_decode_dyn(RZ_NULLABLE const char *src, st64 len) {
+	rz_return_val_if_fail(src, NULL);
+
+	len = calculate_src_length(src, len);
+	if (len < 0 || (len % 2) != 0) {
```

Comment:
```suggestion
	if (len < 0 || len & 1) {
```

What happens if len is zero?

Reply:
update in 4833c01ef6